### PR TITLE
Add map render profiler

### PR DESCRIPTION
**Pull request by nyalldawson**
This extends the debugging/development tools profiler with a new "map render" profile ( to accompany the existing Startup and Project Load profiles).

Through this tool users are able to identify pain points in their map rendering and track down exactly which layers are causing long map redraws:

image

### CODE REVIEW

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 1]
```diff
@@ -33,19 +33,26 @@
 #include "qgsmessagelog.h"
 #include "qgsmapclippingutils.h"
 #include "qgspointcloudblockrequest.h"
+#include "qgsruntimeprofiler.h"
+#include "qgsapplication.h"
 
 QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *layer, QgsRenderContext &context )
   : QgsMapLayerRenderer( layer->id(), &context )
   , mLayer( layer )
+  , mLayerName( layer->name() )
```
Comment (nirvn) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:42]: Why the need to add a mLayerName member when you store the layer pointer too? Are you hoping to get rid of the layer pointer in the near future?

File: src/core/vector/qgsvectorlayerrenderer.cpp [hunk 2]
```diff
@@ -48,10 +50,14 @@ QgsVectorLayerRenderer::QgsVectorLayerRenderer( QgsVectorLayer *layer, QgsRender
   : QgsMapLayerRenderer( layer->id(), &context )
   , mFeedback( std::make_unique< QgsFeedback >() )
   , mLayer( layer )
+  , mLayerName( layer->name() )
```
Comment (nirvn) [inline src/core/vector/qgsvectorlayerrenderer.cpp:53]: Same question as above re storing both a layer pointer and a layer name.

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -1064,6 +1064,8 @@ QgisApp::QgisApp( QSplashScreen *splash, bool restorePlugins, bool skipBadLayers
     settings.setValue( QStringLiteral( "qgis/main_canvas_preview_jobs" ), true );
   }
   mMapCanvas->setPreviewJobsEnabled( settings.value( QStringLiteral( "qgis/main_canvas_preview_jobs" ), true ).toBool() );
+  // record profiling time on the main canvas only
+  mMapCanvas->mapSettings().setFlag( Qgis::MapSettingsFlag::RecordProfile );
```
Comment (nirvn) [inline src/app/qgisapp.cpp:1068]: Out of curiosity, is there a cost associated with this profiling, and if so do we want to allow for it to be toggled on / off?

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:42]: yeah -- it's not safe to use at all (outside of the constructor). I'm just minimising existing bad practice

File: src/app/qgisapp.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/app/qgisapp.cpp:1068]: It should be negligible -- at most each layer render will trigger a handful of signals and some cheap QElapsedTimer calculations. 

The main reason for limiting it to the map canvas only is that I don't want results shown for other canvas renders (eg the CRS selector extent preview widget) .

**Review by nirvn — APPROVED** [review]

**Pull request merged**